viewer/src/Main.rs

```diff
@@ -0,0 +1,324 @@
+                        match keycode {
+                            Keycode::Q => {
+                                self.scale = self
+                                    .scale
+                                    .saturating_add((keycount * keycount).try_into().unwrap())
+                            } //plus
+                            Keycode::W => {
+                                self.scale = self
+                                    .scale
+                                    .saturating_sub((keycount * keycount).try_into().unwrap())
+                            } //minus
+                            Keycode::E => {
+                                self.scale =
+                                    (self.max_stop - self.min_start) / (self.window_width as u64)
+                            } //reset
+                            Keycode::A => self.scroll = self.scroll.saturating_add(keycount),
+                            Keycode::S => self.scroll = self.scroll.saturating_sub(keycount),
+                            Keycode::D => self.scroll = 0,
+                            _ => {}
+                        }
+                    }
+                    Event::MouseButtonDown { x, y, .. } => {
+                        for zone in &self.draw_zones {
+                            if zone.x_start < x.try_into().unwrap()
+                                && zone.x_stop > x.try_into().unwrap()
+                                && zone.y_start < y.try_into().unwrap()
+                                && zone.y_stop > y.try_into().unwrap()
+                            {
+                                draw_x = x;
+                                draw_y = y;
+                                draw_tag = zone.tag_data.0;
+                                draw_len = zone.tag_data.2 - zone.tag_data.1;
+                            }
+                        }
+                    }
+                    Event::MouseButtonUp { .. } => {
+                        draw_x = 0;
+                        draw_y = 0;
+                    }
+                    _ => {}
+                }
+            }
+            if prev_keycount == keycount {
+                keycount = 0;
+            }
+            prev_keycount = keycount;
+
+            self.canvas.set_draw_color(self.background_color);
+            self.canvas.clear();
+            self.draw_zones.clear();
+            for span in &spans {
+                self.draw_span(span);
+            }
+            if (draw_x > 0) && (draw_y > 0) {
+                Self::draw_text(
+                    &mut self.canvas,
+                    &self.texture_creator,
+                    &font,
+                    draw_x,
+                    draw_y,
+                    draw_tag.try_into().unwrap(),
+                    draw_len.try_into().unwrap(),
+                )?;
+            }
+            self.canvas.present();
+            thread::sleep(Duration::new(0, 1_000_000_000u32 / 30));
```

Comment:
doesn't need to be this PR, but this might be a place to get some performance back on larger files.  If iterating through the spans takes a significant portion of the time for a "frame" (or longer than it), the sleep could be reduced or skipped.

Reply:
I think I'll leave this for a later PR, since there's a chance that testing a new solution will be a fairly involved process, and it's not necessary at the moment.